import all modules we'll need to clean data

In [339]:
import pandas as pd
import json
import zipfile



find and import a dataset. I chose LFPL Collection data from https://data.louisvilleky.gov/datasets/LOJIC::louisville-metro-ky-library-collection-inventory-/explore
I had to save it to a zip file to upload to github because it's gigantic. 

In [340]:
df_zip = zipfile. ZipFile('lou_metro_library_collection.zip')
df = pd.read_csv(df_zip.open('lou_metro_library_collection.csv'))

In [341]:
df = pd.read_csv('lou_metro_library_collection.csv')

In [342]:
df.shape

(1657555, 11)

In [343]:
df.head()

BibNum                                              Title  \
0   707409  Jeff Immelt and the new GE way : innovation, t...   
1  1341361           McGraw-Hill's American idioms dictionary   
2   707411  Robin rescues dinner : 52 weeks of quick-fix m...   
3  1341362  McGraw-Hill's essential ESL grammar : a handbo...   
4   707411  Robin rescues dinner : 52 weeks of quick-fix m...   

                 Author          ISBN  PublicationYear ItemType  \
0   Magee, David, 1965-  9.780072e+12             2009     Book   
1    Spears, Richard A.  9.780071e+12             2007     Book   
2  Miller, Robin, 1964-  9.780307e+12             2009     Book   
3         Lester, Mark.  9.780071e+12             2008     Book   
4  Miller, Robin, 1964-  9.780307e+12             2009     Book   

      ItemCollection   ItemLocation  ItemPrice              ReportDate  \
0  Adult Non-Fiction           Main      25.95  2022/07/01 04:00:00+00   
1     ELL Collection  South Central      20.00  2022/07/01 04:00:00+00   
2  Adult Non-Fiction      Southwest      19.99  2022/07/01 04:00:00+00   
3     ELL Collection  South Central      20.00  2022/07/01 04:00:00+00   
4  Adult Non-Fiction      Southwest      19.99  2022/07/01 04:00:00+00   

   ObjectId  
0         1  
1         2  
2         3  
3         4  
4         5

In [344]:
df.columns

Index(['BibNum', 'Title', 'Author', 'ISBN', 'PublicationYear', 'ItemType',
       'ItemCollection', 'ItemLocation', 'ItemPrice', 'ReportDate',
       'ObjectId'],
      dtype='object')

The ISBN column is throwing Excel number errors. I hate these in Excel format. What dype is it? How can I fix this? 

In [345]:
df['ISBN'].dtypes

dtype('float64')

Looks like it's an object type. Can I convert the ISBN column to an integer? (The answer is no.) Let's try converting to a string. 

In [346]:
df['ISBN'] = df['ISBN'].astype('str')

In [347]:
df.head()

BibNum                                              Title  \
0   707409  Jeff Immelt and the new GE way : innovation, t...   
1  1341361           McGraw-Hill's American idioms dictionary   
2   707411  Robin rescues dinner : 52 weeks of quick-fix m...   
3  1341362  McGraw-Hill's essential ESL grammar : a handbo...   
4   707411  Robin rescues dinner : 52 weeks of quick-fix m...   

                 Author             ISBN  PublicationYear ItemType  \
0   Magee, David, 1965-  9780071605878.0             2009     Book   
1    Spears, Richard A.  9780071478939.0             2007     Book   
2  Miller, Robin, 1964-  9780307451408.0             2009     Book   
3         Lester, Mark.  9780071496421.0             2008     Book   
4  Miller, Robin, 1964-  9780307451408.0             2009     Book   

      ItemCollection   ItemLocation  ItemPrice              ReportDate  \
0  Adult Non-Fiction           Main      25.95  2022/07/01 04:00:00+00   
1     ELL Collection  South Central      20.00  2022/07/01 04:00:00+00   
2  Adult Non-Fiction      Southwest      19.99  2022/07/01 04:00:00+00   
3     ELL Collection  South Central      20.00  2022/07/01 04:00:00+00   
4  Adult Non-Fiction      Southwest      19.99  2022/07/01 04:00:00+00   

   ObjectId  
0         1  
1         2  
2         3  
3         4  
4         5

That works. But now I have a trailing decimal point. How do I eliminate that? Can I strip the trailing .0?

In [348]:
df['ISBN'].str.strip('.0')

0          9780071605878
1          9780071478939
2          9780307451408
3          9780071496421
4          9780307451408
               ...      
1657550    9781982168056
1657551    9781982168056
1657552    9781982168056
1657553    9781982168056
1657554    9780692619803
Name: ISBN, Length: 1657555, dtype: object

That works and now the numbers look like actual ISBN numbers again. Now I need to save this new data inplace somehow? 

In [349]:
df['ISBN'] = df['ISBN'].str.strip('.0')

In [350]:
df.head()

BibNum                                              Title  \
0   707409  Jeff Immelt and the new GE way : innovation, t...   
1  1341361           McGraw-Hill's American idioms dictionary   
2   707411  Robin rescues dinner : 52 weeks of quick-fix m...   
3  1341362  McGraw-Hill's essential ESL grammar : a handbo...   
4   707411  Robin rescues dinner : 52 weeks of quick-fix m...   

                 Author           ISBN  PublicationYear ItemType  \
0   Magee, David, 1965-  9780071605878             2009     Book   
1    Spears, Richard A.  9780071478939             2007     Book   
2  Miller, Robin, 1964-  9780307451408             2009     Book   
3         Lester, Mark.  9780071496421             2008     Book   
4  Miller, Robin, 1964-  9780307451408             2009     Book   

      ItemCollection   ItemLocation  ItemPrice              ReportDate  \
0  Adult Non-Fiction           Main      25.95  2022/07/01 04:00:00+00   
1     ELL Collection  South Central      20.00  2022/07/01 04:00:00+00   
2  Adult Non-Fiction      Southwest      19.99  2022/07/01 04:00:00+00   
3     ELL Collection  South Central      20.00  2022/07/01 04:00:00+00   
4  Adult Non-Fiction      Southwest      19.99  2022/07/01 04:00:00+00   

   ObjectId  
0         1  
1         2  
2         3  
3         4  
4         5

What else do we need to clean up in this dataset? How many unique ItemCollections are there? Do those need help?

In [351]:
df.ItemCollection.unique()

array(['Adult Non-Fiction', 'ELL Collection', 'Adult Fiction', 'Mystery',
       'Adult Audiobook', 'Adult DVD', 'Electronic', 'Music',
       "Children's Picture Book", 'Science Fiction',
       'Book Discussion Kit', 'Older Teen Fiction',
       'Younger Teen  Fiction', 'Adult Paperback', "Children's Fiction",
       'Western', "Children's Picture Paperback", "Children's Paperback",
       'International Collection', 'Teen Non-Fiction',
       "Children's Non-Fiction", 'Kentucky History', 'Natural Resources',
       'Oversize', 'Holiday', 'Urban Fiction', 'Bestsellers',
       "Children's Board Book", 'Storytime Collection',
       'Preschool  Picture Book', "Children's Easy Reader", 'Readalongs',
       "Children's DVD", "Children's Reference", 'Adult Reference',
       'Interlibrary Loan', "Children's Music", nan,
       'Adult Paperbacks Tall', "Children's Easy Reader Paperback",
       'Large Print', 'Caldecott/Newbery', "Children's Audiobook",
       'OlderTeen Audiobook', 'Youn

There is one "nan" value in here. Let's find it and destroy it?

In [352]:
df2 = df.loc[df['ItemCollection'] == 'nan']

In [353]:
print(df2)

Empty DataFrame
Columns: [BibNum, Title, Author, ISBN, PublicationYear, ItemType, ItemCollection, ItemLocation, ItemPrice, ReportDate, ObjectId]
Index: []


Okay so maybe that was a placeholder because that query returns an empty dataframe. This is actually a really clean dataset so far. I approve of LFPL. One thing that might make it easier to work with is cleaning up the ItemPrice column. I can round those to the nearest dollar since we have values like 19.99 and 20.00. Using 20.00 across the board would make things a lot more succinct. Let's do that next. I can use the ceiling function here. 

In [354]:
df['ItemPrice'] = df['ItemPrice'].apply(np.ceil)

In [355]:
df.head()

BibNum                                              Title  \
0   707409  Jeff Immelt and the new GE way : innovation, t...   
1  1341361           McGraw-Hill's American idioms dictionary   
2   707411  Robin rescues dinner : 52 weeks of quick-fix m...   
3  1341362  McGraw-Hill's essential ESL grammar : a handbo...   
4   707411  Robin rescues dinner : 52 weeks of quick-fix m...   

                 Author           ISBN  PublicationYear ItemType  \
0   Magee, David, 1965-  9780071605878             2009     Book   
1    Spears, Richard A.  9780071478939             2007     Book   
2  Miller, Robin, 1964-  9780307451408             2009     Book   
3         Lester, Mark.  9780071496421             2008     Book   
4  Miller, Robin, 1964-  9780307451408             2009     Book   

      ItemCollection   ItemLocation  ItemPrice              ReportDate  \
0  Adult Non-Fiction           Main       26.0  2022/07/01 04:00:00+00   
1     ELL Collection  South Central       20.0  2022/07/01 04:00:00+00   
2  Adult Non-Fiction      Southwest       20.0  2022/07/01 04:00:00+00   
3     ELL Collection  South Central       20.0  2022/07/01 04:00:00+00   
4  Adult Non-Fiction      Southwest       20.0  2022/07/01 04:00:00+00   

   ObjectId  
0         1  
1         2  
2         3  
3         4  
4         5

It looks like the prices are now rounded up to 20.0 instead of 19.99, which should be a little easier to work with. I think that makes the cleaning element number 2 and this dataset is actually really well maintained so there won't be much more to edit/clean here. But it might be fun to work with in the future. 